### PR TITLE
Addition of Landmark Storage Functionality 

### PR DESCRIPTION
**Pull request by artemmikhalitsin**
Addresses #100 and #101. 

Landmarks are now stored and retrieved. Saved landmarks are checked upon loading. 

The test coverage is somewhat limited for tasks #99, #100 and #101, this is due to the fact that the additions made for them were aesthetic in nature (#99) or db handling augmentations to the implementation of #26. An espresso test was added for the checkmarks (which indicate selected landmarks) and a unit test introduced for the updated trip database helper class (a key class for storage of trips and hence landmarks).

### CODE REVIEW

File: app/src/main/java/org/wikipedia/travel/landmarkpicker/LandmarkFragment.java [hunk 1]
```diff
@@ -136,22 +137,42 @@ private void onLoadPage(@NonNull PageTitle title, HistoryEntry entry) {
...
+        return null;
     }
 
-    //Saves list of selectedLandmarks
-    private void onSave(List<LandmarkCard> saveList){
-        Callback callback = getCallback();
-        if (callback != null) {
-            callback.onSave(saveList);
+    private void loadSelectedCards() {
+        if(this.destination != null) {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/travel/landmarkpicker/LandmarkFragment.java]: `getCallback()` should ideally only be invoked once, so it should be initially stored in a variable before the conditional, and every instance of `getCallback()` in this function should use that variable

File: app/src/main/java/org/wikipedia/travel/database/TripDbHelper.java [hunk 2]
```diff
@@ -114,6 +117,58 @@ public Trip getFullListById(long id) {
         return list;
     }
 
+    public Void addUserLandmarks(long tripId, List<LandmarkCard> landmarks) {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/travel/database/TripDbHelper.java]: Is the reason as to why it's `Void` vs `void` because of the CallbackTask?

File: app/src/main/java/org/wikipedia/travel/landmarkpicker/LandmarkFragment.java [hunk 3]
```diff
@@ -311,11 +331,15 @@ public void onClick(View v) {
                         //Adds the LandmarkCard object, comprised of card title and thumbUrl into list
                         if (card.getChecked()) {
                             card.setChecked(false);
-                            selectedLandmarks.remove(card);
+                            if(getCallback() != null) {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/travel/landmarkpicker/LandmarkFragment.java]: Same comment applies about storing `getCallback()` in a `Callback` variable

**Review by amawai — CHANGES_REQUESTED** [review]
This pull request fully implements the feature of saving landmarks with relevant tests. Moreover, database functionality is consistent with the rest of the application. The only thing to change is storing the `getCallback()` in a variable to reduce the number of method calls.

File: app/src/main/java/org/wikipedia/travel/landmarkpicker/LandmarkFragment.java (hunk 1 above)
Comment (artemmikhalitsin) [reply app/src/main/java/org/wikipedia/travel/landmarkpicker/LandmarkFragment.java]: I technically agree, however, this is the style that the previous devs used. See HistoryFragment.java for example.

File: app/src/main/java/org/wikipedia/travel/database/TripDbHelper.java (hunk 2 above)
Comment (artemmikhalitsin) [reply app/src/main/java/org/wikipedia/travel/database/TripDbHelper.java]: Yeah exactly. Previously I set it as Object, but i noticed the devs using Void instead. I guess it's more comprehensive.

**Review by amawai — APPROVED** [review]

File: app/src/main/java/org/wikipedia/database/contract/UserLandmarkContract.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Created by Artem on 2018-04-12.
+ */
+
+public interface UserLandmarkContract {
+    String TABLE = "usertriplandmarks";
+    Uri URI = Uri.withAppendedPath(AppContentProviderContract.AUTHORITY_BASE, "/localtriplandmark");
+
+
+    interface Col {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/database/contract/UserLandmarkContract.java]: Codacy Issue found: [interfaces should describe a type and hence have methods.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541239)

File: app/src/main/java/org/wikipedia/database/contract/UserLandmarkContract.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import org.wikipedia.database.column.IdColumn;
+import org.wikipedia.database.column.LongColumn;
+import org.wikipedia.database.column.StrColumn;
+
+/**
+ * Created by Artem on 2018-04-12.
+ */
+
+public interface UserLandmarkContract {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/database/contract/UserLandmarkContract.java]: Codacy Issue found: [interfaces should describe a type and hence have methods.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541239)

**Comment by amawai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541239&bid=7051448)

**Review by jadmalek — APPROVED** [review]
This was an extremely important addition to the trip planner feature and essentially constitutes the basis for said feature, with the user now being able to select and save landmarks for a trip. While there could have been more comments to further describe the nuances of the database interactions/augmentations and more unit test coverage could have been integrated, the code here is very nicely implemented, follows established coding rules, and integrates well with previously composed components of the trip planner from the previous release. The unit test does cover aspects (including edge cases) of the db interaction and is hence adequate for said purpose, further adding to the test coverage added by the team for US #22, #26 and the task #56.

File: app/src/main/java/org/wikipedia/travel/MainPlannerFragment.java [hunk 6]
```diff
@@ -239,11 +261,20 @@ public boolean onBackPressed() {
...
+            TripDbHelper.instance().addUserLandmarks(openTrip.getId(), addedLandmarks);
+            TripDbHelper.instance().deleteUserLandmarks(openTrip.getId(), removedLandmarks);
+            return null;
+        }, new CallbackTask.DefaultCallback<Void>() {
             @Override
-            public void success(Object o) {
+            public void success(Void o) {
                 super.success(o);
+                //Reset lists which track added and removed cards
+                addedLandmarks.clear();
```
[the remark this answers is not in the document]
Comment (jadmalek) [reply app/src/main/java/org/wikipedia/travel/MainPlannerFragment.java]: Clarified by @artemmikhalitsin: the clear() method needs to be called by lists in order to avoid duplicates when the list is regenerated for view in the landmarks screen.

File: app/src/main/java/org/wikipedia/travel/database/UserLandmarkTable.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import org.wikipedia.database.column.Column;
+import org.wikipedia.database.contract.UserLandmarkContract;
+
+import java.util.ArrayList;
+
+/**
+ * Created by Artem on 2018-04-12.
+ */
+
+public class UserLandmarkTable extends DatabaseTable<UserLandmark> {
```
[the remark this answers is not in the document]
Comment (jadmalek) [reply app/src/main/java/org/wikipedia/travel/database/UserLandmarkTable.java]: Clarified by @artemmikhalitsin and @sophiaquach16: landmarks are stored as a list of string values (corresponding to the cards in question) in one column and its associated trip id is the id for said values. The retrieval essentially takes place in the main planner and landmark fragments, with the respective methods included in both responsible for enabling the saving and retrieval of a trip's selected landmarks.

File: app/src/androidTest/java/org/wikipedia/travel/database/DestinationDbHelperTest.java [hunk 8]
```diff
@@ -1,95 +1,97 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static junit.framework.Assert.assertNotNull;
+import static junit.framework.Assert.assertTrue;
+import static org.junit.Assert.assertEquals;
+
+public class DestinationDbHelperTest {
+
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/travel/database/DestinationDbHelperTest.java]: Codacy Issue found: [Line has trailing spaces.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541239)

**Comment by amawai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541239&bid=7051448)

**Comment by jadmalek** [comment]
**Acceptance Testing**

| Test #         | Inputs | Expected Output  | Status |
| ------------------ | ------------- | ------------- | ------------- |
| 1. Plan a Trip card appears on application's main page.  | Wikipedia application launched.  | A card for the trip planner is displayed, along with the other cards, on the main page.  | PASS  |
| 2. Plan a Trip page display.  | Press on Plan a Trip arrow icon on card.  | The Plan a Trip page is depicted, including a list of trips and the option to plan a new trip.   | PASS  |
| 3. Destination picker display.  | Press on the "+ Plan a new trip" button. | Destinations page is depicted, including a search bar for a destination to be inputted.   | PASS  |
| 4. Particular destination selection.  | Type in the name of the destination and select on the dropdown menu, or simply click on the dropdown menu and select a desired destination. | Selected destination will appear in text format above the search bar.   | PASS  |
| 5. Particular destination selection.  | Type in the name of the destination and select on the dropdown menu, or simply click on the dropdown menu and select a desired destination. | Selected destination will appear in text format above the search bar.   | PASS  |
| 6. Particular destination selection.  | Pick a date from the calendar that pops up upon clicking the "Select Date" button. | Selected date will appear above the "Select Date" button.   | PASS  |
| 7. Landmarks page display.  | Select the next (arrow) icon. | Landmarks for the selected destination appear, where each landmark has a checkbox next to it.   | PASS  |
| 8. View landmarks articles.  | Click on any of the landmark. | Article page corresponding to that landmark is displayed.   | PASS  |
| 9. Save landmarks.  | Check off boxes of desired landmarks. | "Trip has been saved" message is displayed.   | PASS  |
| 9. Check that saved landmarks are indicated as such in landmarks screen.  | Select on edit ("pencil") icon to modify a saved planned trip and select on next button in each of the screens of the trip planner until the landmarks screen is displayed. | Previously saved landmarks will have checks in the corresponding checkboxes. | PASS  |


**Demo Steps**

1. Launch the application.
2. Scroll down the main page until the Plan a Trip card is visible.
3. Press on the Plan a trip arrow icon to view the Plan a Trip screen.
4. Press on the "+ Plan a new trip" button in order to begin planning a new trip.
5. Enter a particular destination.
6. Press the next button.
7. Pick a particular Departure Date from the "Select Date" calendar.
8. Press the next button.
9. Click on a landmark.
10. Press back to return to the destination's Landmarks screen.
11. Select landmark checkboxes.
12. Select the Save button.
13. Press back until on main page of application.
14. Select on the edit ("pencil") of most recently planned trip.
15. Press next button on destination and date screens to get to the landmarks screen.
16. Confirm that the selected landmarks are checked off, indicating that these are the desired landmarks for the trip.
